src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Instrument.cs

```diff
@@ -88,6 +89,17 @@ protected void Publish()
                 return;
             }
 
+            // MeterListener has a static constructor that creates runtime metrics instruments.
+            // We need to ensure this static constructor is called before starting to publish the instrument.
+            // This is necessary because creating runtime metrics instruments will cause re-entry to the Publish method,
+            // potentially resulting in a deadlock due to the SyncObject lock.
+            // Sequence of the deadlock:
+            //   1. An application creates an early instrument (e.g., Counter) before the MeterListener static constructor is executed.
+            //   2. Instrument.Publish is called and enters the SyncObject lock.
+            //   3. Within the lock block, MeterListener is called, triggering its static constructor.
+            //   4. The static constructor creates runtime metrics instruments, causing re-entry to Instrument.Publish and leading to a deadlock.
+            RuntimeHelpers.RunClassConstructor(typeof(MeterListener).TypeHandle);
```

Comment:
I've been thinking about it and this seems OK - I didn't come up with another place where the MeterListener static constructor would be a problem. In the future if we did discover any more it might be worthwhile to move the runtime metrics initialization out of the static constructor and instead do it as a lazy initialization within the MeterListener instance constructor.

Reply:
Yes, we can consider moving the initialization later out of the static constructor.